ecosystem/python/sdk/aptos_sdk/bcs.py

```diff
@@ -0,0 +1,341 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Aptos
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+This is a simple BCS serializer and deserializer. Learn more at https://github.com/diem/bcs
+"""
+
+from __future__ import annotations
+
+import io
+import typing
+import unittest
+
+MAX_U8 = 2**8 - 1
+MAX_U16 = 2**16 - 1
+MAX_U32 = 2**32 - 1
+MAX_U64 = 2**64 - 1
+MAX_U128 = 2**128 - 1
+
+
+class Deserializer:
+    _input: io.BytesIO
+    _length: int
+
+    def __init__(self, data: bytes):
+        self._length = len(data)
+        self._input = io.BytesIO(data)
+
+    def remaining(self) -> int:
+        return self._length - self._input.tell()
+
+    def bool(self) -> bool:
+        value = int.from_bytes(self._read(1), byteorder="little", signed=False)
+        if value == 0:
+            return False
+        elif value == 1:
+            return True
+        else:
+            raise Exception("Unexpected boolean value: ", value)
+
+    def bytes(self) -> bytes:
+        return self._read(self.uleb128())
+
+    def fixed_bytes(self, length: int) -> bytes:
+        return self._read(length)
+
+    def map(
+        self,
+        key_decoder: typing.Callable[[Deserializer], typing.Any],
+        value_decoder: typing.Callable[[Deserializer], typing.Any],
+    ) -> Dict[typing.Any, typing.Any]:
+        length = self.uleb128()
+        values = {}
+        while len(values) < length:
+            key = key_decoder(self)
+            value = value_decoder(self)
+            values[key] = value
+        return values
+
+    def sequence(
+        self,
+        value_decoder: typing.Callable[[Deserializer], typing.Any],
+    ) -> List[typing.Any]:
+        length = self.uleb128()
+        values = []
+        while len(values) < length:
+            values.append(value_decoder(self))
+        return values
+
+    def str(self) -> str:
+        return self.bytes().decode()
+
+    def struct(self, struct: typing.Any) -> typing.Any:
+        return struct.deserialize(self)
+
+    def u8(self) -> int:
+        return self._read_int(1)
+
+    def u16(self) -> int:
+        return self._read_int(2)
+
+    def u32(self) -> int:
+        return self._read_int(4)
+
+    def u64(self) -> int:
+        return self._read_int(8)
+
+    def uleb128(self) -> int:
+        value = 0
+        shift = 0
+
+        while value <= MAX_U128:
```

Comment:
This is also a bit of interesting. When you check MAX_U128 here, that gives people a feeling that uleb128 is bounded at MAX_U128. uleb128 just means, it is encoded as base128. In BCS, uleb128 is used only for small integers, such as the length of sequences. If you really want to add a bound check, we should bound this with MAX_U32 not MAX_U128. 

See implementation in Rust BCS:
```
#[allow(clippy::integer_arithmetic)]
    fn parse_u32_from_uleb128(&mut self) -> Result<u32> {
        let mut value: u64 = 0;
        for shift in (0..32).step_by(7) {
            let byte = self.next()?;
            let digit = byte & 0x7f;
            value |= u64::from(digit) << shift;
            // If the highest bit of `byte` is 0, return the final value.
            if digit == byte {
                if shift > 0 && digit == 0 {
                    // We only accept canonical ULEB128 encodings, therefore the
                    // heaviest (and last) base-128 digit must be non-zero.
                    return Err(Error::NonCanonicalUleb128Encoding);
                }
                // Decoded integer must not overflow.
                return u32::try_from(value)
                    .map_err(|_| Error::IntegerOverflowDuringUleb128Decoding);
            }
        }
        // Decoded integer must not overflow.
        Err(Error::IntegerOverflowDuringUleb128Decoding)
    }
```

source code https://docs.rs/crate/bcs/latest/source/src/de.rs

Reply:
good call, the code diverged quite a bit from what I had remembered :)

so uleb128 is only used for internal data and u128 is external... need to update bcs to be smart like that!!